src/Compilers/Core/Portable/Text/ChangedText.cs

```diff
@@ -240,149 +240,187 @@ private static ImmutableArray<TextChangeRange> Merge(IReadOnlyList<ImmutableArra
-                            goto nextOldChange;
-                        }
+
+                        // adjust the new change as follows:
+                        // 1. add the old change's deletion to its deletion
+                        // 2. subtract the old change's insertion from its deletion
+                        // 3. add oldDelta to its start position so that its adjusted start position is the same as the oldChange that we removed
+                        //    - note that this can result in the newChange having a negative Start, but this gets adjusted back into a non-negative Start in the next iteration.
+                        var newDeletion = newChange.Span.Length + oldChange.Span.Length - oldChange.NewLength;
+                        newChange = new TextChangeRange(new TextSpan(oldChange.Span.Start + oldDelta, newDeletion, validateStart: false), newChange.NewLength);
+                        if (tryGetNext(oldChanges, ref oldIndex, out oldChange)) continue;
+                        else break;
                     }
                 }
-                else
-                {
-                    // no more old changes, just add adjusted new change
-                    var adjustedNewChange = new TextChangeRange(new TextSpan(newChange.Span.Start - oldDelta, newChange.Span.Length), newChange.NewLength);
-                    AddRange(list, adjustedNewChange);
-                    newIndex++;
-                    goto nextNewChange;
-                }
             }
-            else
+
+            // there may be remaining old changes or remaining new changes (not both)
+            Debug.Assert(oldIndex == oldChanges.Length || newIndex == newChanges.Length);
+
+            while (oldIndex < oldChanges.Length)
             {
-                // no more new changes, just add remaining old changes
-                while (oldIndex < oldChanges.Length)
-                {
-                    AddRange(list, oldChanges[oldIndex]);
-                    oldIndex++;
-                }
+                add(oldChange); // no need to adjust oldDelta at this point
+                tryGetNext(oldChanges, ref oldIndex, out oldChange);
             }
 
-            return list.ToImmutableArray();
-        }
+            while (newIndex < newChanges.Length)
+            {
+                addAdjustedNewChange(newChange);
+                tryGetNext(newChanges, ref newIndex, out newChange);
+            }
 
-        private static void AddRange(List<TextChangeRange> list, TextChangeRange range)
-        {
-            if (list.Count > 0)
+            return builder.ToImmutableAndFree();
+
+            static bool tryGetNext(ImmutableArray<TextChangeRange> changes, ref int index, out TextChangeRange next)
             {
-                var last = list[list.Count - 1];
-                if (last.Span.End == range.Span.Start)
+                if (++index < changes.Length)
```

Comment:
```suggestion
                ++index;
                if (index < changes.Length)
```

having these be separate means tehre's no need to even consider order of operations/precdence/etc.  while yoru code is 100% fine as per the rules of C# it adds a needless amount of mental overhead in an algorithm that is already quite complex (and clearly error prone).  We should err toward clarity and simplicity here unless there is a deep need to do otherwise.

Reply:
Done